mds: wait snapupdate to finish before notifying to clients

xiubli opened the pull request:
When making snapshot the MDS will notify the mds first and then the
clients, but the other MDSes may get the notify late and just after
the clients flush the snapcap to them. The MDSes will just drop the
snapcap flush request to floor.

We need to wait for a while and make sure all the other MDSes get
notifications first.

The same with rmsnap and renamesnap ops.

Fixes: https://tracker.ceph.com/issues/56011
Signed-off-by: Xiubo Li <xiubli@redhat.com>





<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

xiubli committed:
mds: wait snapupdate to finish before notifying to clients

When creating snapshot, the primary MDS will notify other mds first
and then the clients, but the other MDSes may get the notify late and
just after the clients flush the snapcap to them. The other MDSs will
just drop the snapcap flush request to the floor.

We need to wait for a while and make sure all the other MDSes get
notifications first.

The same with rmsnap and renamesnap ops.

Fixes: https://tracker.ceph.com/issues/56011
Signed-off-by: Xiubo Li <xiubli@redhat.com>

xiubli committed:
qa: increase the test number to 100 for snaptest-snap-rm-cmp.sh

For some potential bugs it's hard to be reproduced and we need
to test more. Locally testing 100 times will take less than 5
minutes.

Fixes: https://tracker.ceph.com/issues/56011
Signed-off-by: Xiubo Li <xiubli@redhat.com>

mchangir reviewed (approved):
LGTM

lxbsz commented:
Resolved the conflicts.

kotreshhr commented on src/mds/MDCache.cc:
```diff
@@ -10051,6 +10061,29 @@ void MDCache::handle_snap_update(const cref_t<MMDSSnapUpdate> &m)
+  CInode *in = get_inode(m->get_ino());
+  if (in) {
+    in->snap_update_ref--;
+    if (!in->snap_update_ref) {
```
What happens when one of the snap_update_reply is not received because the mds might have gone down before replying ?

lxbsz replied:
It will wait infinitely here I think. Maybe we should set a timetout here ? Any better idea ?

vshankar replied:
Maybe remove the mds from the set when _this_ MDS receives an mdsmap that has the peer mds missing?

vshankar replied:
@lxbsz ping?

lxbsz replied:
Sorry, I missed this. Will try to fix it. Thanks @kotreshhr @vshankar

lxbsz replied:
> Maybe remove the mds from the set when _this_ MDS receives an mdsmap that has the peer mds missing?

This sounds good to me and appended a commit to fix this.

vshankar commented:
jenkins test api

xiubli committed:
mds: fix possible ref leakage bug

Just in case any of the replica MDSs is down and won't get a chance
to reply us.

Fixes: https://tracker.ceph.com/issues/56011
Signed-off-by: Xiubo Li <xiubli@redhat.com>

lxbsz commented:
jenkins test make check

dparmar18 commented:
@lxbsz this needs a rebase

lxbsz commented:
> @lxbsz this needs a rebase

Done. Thanks for reminding.

mchangir commented:
> When making snapshot the MDS will notify the mds first and then the clients, but the other MDSes may get the notify late and just after the clients flush the snapcap to them. The MDSes will just drop the snapcap flush request to floor.

I think "When **creating** snapshot, the **primary** MDS will notify **other** mds first ..." would sound better
also
"The **other** MDSs will just drop the snapcap flush request to the floor"
> 
> We need to wait for a while and make sure all the other MDSes get notifications first.
> 
> The same with rmsnap and renamesnap ops.

lxbsz commented:
> > When making snapshot the MDS will notify the mds first and then the clients, but the other MDSes may get the notify late and just after the clients flush the snapcap to them. The MDSes will just drop the snapcap flush request to floor.
> 
> I think "When **creating** snapshot, the **primary** MDS will notify **other** mds first ..." would sound better also "The **other** MDSs will just drop the snapcap flush request to the floor"
> 
> > We need to wait for a while and make sure all the other MDSes get notifications first.
> > The same with rmsnap and renamesnap ops.

Revised it and thanks @mchangir

mchangir commented on src/mds/CInode.h:
```diff
@@ -1118,7 +1118,7 @@ class CInode : public MDSCacheObject, public InodeStoreBase, public Counter<CIno
   client_t loner_cap = -1, want_loner_cap = -1;
 
-  uint32_t snap_update_ref = 0;
+  std::set<mds_rank_t> snap_update_ref_set;
```
can we use `std::unordered_set<>` instead for O(1) access ?
though there won't be millions of replicas to cause performance problems using `std::set<>`

lxbsz replied:
Yeah, this looks good to me. I will revise this. Thanks @mchangir

gregsfortytwo replied:
@chrisphoffman just pointed me at this PR and I noticed this comment. unordered_hash is a hash map and they are quite large, so it is VERY BAD to put those in to MDSCacheObjects.
How much space does an unordered_set take up? We do want to pay some attention to their size given we have hundreds of thousands or millions of them in the cache...

vshankar replied:
Given that this holds the peer mds rank for which snap update has been sent, it can be a 8bit value (char) which can accommodate MAX_MDS.

lxbsz commented:
jenkins retest this please

vshankar commented:
@lxbsz Please rebase and push - I'll try to have a look again.

vshankar commented:
@lxbsz please rebase. Will put this to test.

lxbsz commented:
> @lxbsz please rebase. Will put this to test.

Sorry @vshankar I missed you last two comments about this. Rebased it now.

lxbsz commented:
jenkins retest this please

dparmar18 commented:
@vshankar are we still looking at some of these PRs?

vshankar commented:
> @vshankar are we still looking at some of these PRs?

Yes. I had this with me, but unable to get to it.
